lib/sycamore/sycamore/data/table.py

```diff
@@ -0,0 +1,187 @@
+from collections import OrderedDict
+from dataclasses import dataclass, field
+from typing import Any, Optional, TypeVar, Union
+import xml.etree.ElementTree as ET
+
+from sycamore.data import BoundingBox
+from PIL import Image, ImageDraw
+import numpy as np
+from pandas import DataFrame
+
+
+# This is part of itertools in 3.10+.
+# Adding here to support 3.9
+def _pairwise(iterable):
+    iterator = iter(iterable)
+    a = next(iterator, None)
+    for b in iterator:
+        yield a, b
+        a = b
+
+
+# This data model is similar to that used by Textract and TableTransformers.
+# It is intended to be a common intermediate representation for a variety of
+# table structure recognition models.
+@dataclass(frozen=True)
+class TableCell:
+    content: str
+    rows: list[int]
+    cols: list[int]
+    is_header: bool = False
+    bbox: Optional[BoundingBox] = None
+    # Model/format specific properties
+    properties: dict[str, Any] = field(default_factory=dict)
+
+    def __post_init__(self):
+        # Validate that row and column spans are contiguous.
+        for a, b in _pairwise(self.rows):
+            if a + 1 != b:
+                raise ValueError(f"Found non-contiguous rows in {self}.")
+
+        for a, b in _pairwise(self.cols):
+            if a + 1 != b:
+                raise ValueError(f"Found non-contiguous cols in {self}.")
+
+
+class Table:
+    def __init__(self, cells: list[TableCell], caption: Optional[str] = None):
+        self.cells: list[TableCell] = sorted(cells, key=lambda tc: (min(tc.rows), min(tc.cols)))
+        self.caption = caption
+        self.num_rows = max(max(c.rows) for c in self.cells) + 1
+        self.num_cols = max(max(c.cols) for c in self.cells) + 1
+
+    def __eq__(self, other):
+        if type(other) is not type(self):
+            return False
+
+        if self.num_rows != other.num_rows or self.num_cols != other.num_cols:
+            return False
+
+        # TODO: Yikes! this is expensive. Do the sorting in init or something.
+        if self.cells != other.cells:
+            return False
+
+        return True
+
+    def __hash__(self):
+        return hash((self.cells))
+
+    # This algorithm is modified from the TableTransformers code. The conversion to Pandas/CSV
+    # is necessarily lossy since these formats requires tables to be square and don't support
+    # cells spanning multiple rows/columns.
+    #
+    # One of the main decisions here is how to render cells that span multiple rows and columns,
+    # since they are not nativey supported in CSV. The standard approaches are (1) to duplicate
```

Comment:
do we need to stick to csv? Can we do json or something and only do csv style conversion when necessary?

Reply:
So in this PR we aren't actually storing anything as CSV -- it's only provided as a target for conversion. The question I guess is what do we embed/store in OpenSearch. I'm open to suggestions for that. My impression was that LLMs/embedding models would probably do better with CSV/HTML since they are common ways to represent tables, but happy to try something else.